Added design document for OMEGA metadata management

philipwjones opened the pull request:
Adds a design document for managing field and global metadata within OMEGA.

cbegeman commented on components/omega/doc/design/Metadata.md:
```diff
@@ -0,0 +1,320 @@
+           // Dimensions - only for defining fields/variables
+           I4 nDims;   // number of dimensions
+           std::vector<MetadataDim*> dims; // pointer to the defined
+                                           // dimension for each dim
```
In MPAS-Ocean to support fill-value masking we had to add a pointer `mask` to `cellMask` or `edgeMask` (and should include other masks such as `landIceMask`). I don't know if you had another way around this in mind or if we need something similar here.

philipwjones replied:
At this level, the fillValue is only specifying the value to be used as a fill so that apps can check for that value in the field. The actual filling and masking will take place elsewhere.

cbegeman commented on components/omega/doc/design/Metadata.md:
```diff
@@ -0,0 +1,320 @@
+             const std::vector<MetadataDims*> // vector of dim pointers
+             );
+
+Do we want to require any other metadata (valid min/max, fill value)?
```
See above comments about fill value masking. 

I think that including valid min/max metadata could be useful, that is, they could be used directly in the new version of `ocn_validate_state` to check whether values are in bounds.

philipwjones replied:
add valid min/max as required metadata. see above comment about fill value and masking

xylar commented on components/omega/doc/design/Metadata.md:
```diff
@@ -0,0 +1,320 @@
+be used to extract all of the metadata for a given type. For
+example, to retrieve all R4 metadata from a defined field:
+
+   std::map<std::string, R4> r4meta = MetadataGetAll(varName);
+   std::string metaName;
+   R4 metaVal
+
+   // use std::map functionality to loop through all metadata
+   for (auto it=r4meta.begin(); it != r4.meta.end(), ++it){
+      metaName = it->first;  // retrieve name part of meta pair
+      metaVal  = it->second; // retrieve value part of meta pair
+
+      // do whatever needed with metadata
+   }
```
Here in particular, the GitHub preview of this document isn't coming out quite right because you have 3 spaces rather than 4.  An easy way to avoid this is to put code blocks in triple-back-quotes.  If you specify `c++`:
````
```c++
```
````
 the code will be colored according to syntax for improved (at least in my view) readability.  Here, that would become:
````suggestion
```c++
std::map<std::string, R4> r4meta = MetadataGetAll(varName);
std::string metaName;
R4 metaVal

// use std::map functionality to loop through all metadata
for (auto it=r4meta.begin(); it != r4.meta.end(), ++it){
   metaName = it->first;  // retrieve name part of meta pair
   metaVal  = it->second; // retrieve value part of meta pair
    // do whatever needed with metadata
}
```
````

xylar commented on components/omega/doc/design/Metadata.md:
```diff
@@ -0,0 +1,320 @@
+
+## 5 Verification and Testing
+
+Requirement 2.3 will be enforced as needed when defining available
+fields so will not be tested as part of unit testing.
```
It could still be a good idea to run a cf-checker either in Omega or in polaris.  It seems like it might be easier to do it on the polaris side (so not as a unit test).  But it seems pretty likely that some CF non-compliance will emerged by mistake in the development process so we do want some way to catch this.

philipwjones replied:
added text for this (though probably will be a polaris test)

xylar reviewed (changes requested):
This looks good!  A few comments and some requested formatting changes to the document itself.

philipwjones commented:
@xylar and @cbegeman  Can you please take another look at this? I've added a group feature to create a shorthand for groups of fields so that the contents in config files don't get too long (sorta like MPAS packages but at the metadata level). I've significantly changed a lot of the implementation details - my previous implementation would not have worked and ChatGPT had some good advice when I asked it for code ;^).  But the functionality and requirements are largely the same - I did eliminate the model config-in-metadata and we'll do provenance in another way. Finally, I changed formatting to comply with the new template and documentation structure.

philipwjones commented on components/omega/doc/design/Metadata.md:
```diff
@@ -0,0 +1,320 @@
+set of required metadata and interfaces must enforce this minimum set.
+This minimum will be defined later, but for variables, this would
+typically include a name (short), units, long name/description,
+standard CF name (if exists), and dimensions.
```
Added a fillValue as a required metadata entry

xylar commented:
@philipwjones, sorry for the delay on this. I'll take a look as soon as I can, hopefully later today or tomorrow.

cbegeman commented on components/omega/doc/design/Metadata.md:
```diff
@@ -0,0 +1,408 @@
+## 4 Design
+
+The general philosophy in this design is to have modules that
+"own" the variable define the field metadata and add it to a
```
```suggestion
"own" the variable, define the field metadata and add it to a
```
Is a comma missing here?

philipwjones replied:
@cbegeman Actually, no.  In this case "that own the variable" was meant to be a restrictive clause modifying "modules", but I can see the point of confusion.  Maybe clearer if I write "The general philosophy is that modules will define the metadata for each field they own and add it..."  Thanks for re-reviewing and I'll make that change after seeing what other changes are suggested.

cbegeman reviewed (approved):
@philipwjones It seems like you've given this a lot of thought and I don't have additional feedback to offer at this point. The metadata groups seem appropriate to me. Thanks!

xylar reviewed (approved):
@philipwjones, this looks excellent.  I reread the document and it all looks great to me.  Thank you for reformatting based on the updated template.

I'm very excited to work with this code.
